### ISSUE TITLE
Old task statuses remaining visible, given high parallelism and small window sizes

### ISSUE THREAD

**Issue by mathjeff** [gradle/gradle#11228]
(The build that I'm working on is the AndroidX build ( https://android.googlesource.com/platform/frameworks/support/+/a19e98dd60f36036525cd992cf8210022e500977/README.md ), but)
I can reproduce this behavior when building Gradle itself:

  ```
  git clone https://github.com/gradle/gradle
  cd gradle
  git checkout v5.6.2
  # Resize the terminal window such that it shows very few rows. I was able to resize my terminal to 5 lines
  ./gradlew --max-workers=100 build

  # Notice that even after 10 minutes, after Gradle has moved beyond building the buildSrc project, the status display still says that Gradle is building buildSrc:
 ```

  <============-> 97% EXECUTING [10m 0s]
  > Building buildSrc > :buildSrc:docs:ktlintMainSourceSetCheck
  > Building buildSrc > :buildSrc:integrationTesting:ktlintMainSourceSetCheck

I haven't completely finished investigated what's going on but I think there are a couple of root causes.

I think one root cause is that

  `attach(operations.start(startEvent.getStatus(), startEvent.getCategory(), startEvent.getProgressOperationId(), startEvent.getParentProgressOperationId()));`

at

  https://github.com/gradle/gradle/blob/7ee00dbffc91c603d010bb10b96523a77880d443/subprojects/logging/src/main/java/org/gradle/internal/logging/console/WorkInProgressRenderer.java#L185

creates a new ProgressOperation object, and if that ProgressOperation object is added into `unassignedProgressOperations` then it may not be removed by

  `unassignedProgressOperations.remove(operation)`

at

  https://github.com/gradle/gradle/blob/7ee00dbffc91c603d010bb10b96523a77880d443/subprojects/logging/src/main/java/org/gradle/internal/logging/console/WorkInProgressRenderer.java#L141

.

I think that a second root cause is that

  `attach(operation.getParent());`

at

  https://github.com/gradle/gradle/blob/7ee00dbffc91c603d010bb10b96523a77880d443/subprojects/logging/src/main/java/org/gradle/internal/logging/console/WorkInProgressRenderer.java#L144

adds the parent to the display when its child terminates but that if the termination message for the parent was already received then nothing remains to terminate the parent, and the parent remains forever.

I could look into submitting a fix if this looks reasonable and if nobody else has a chance.

**Comment by mathjeff** [comment]
Also note that this seems to be happening whenever the console is only able to display less than half as many lines as there are workers in the current build

(the reproduction instruction above about specifically shrinking the window down to 5 lines was just to make it easier to notice)

**Comment by mathjeff** [comment]
Hi stale bot, I encounter this issue fairly often so I looked into it and I think I will be uploading a pull request soon. I don't think that I have permission to reopen it, though; I hope that's not too confusing

**Comment by mathjeff** [comment]
For reference, the true root cause was that https://github.com/gradle/gradle/blob/7ee00dbffc91c603d010bb10b96523a77880d443/subprojects/logging/src/main/java/org/gradle/internal/logging/console/WorkInProgressRenderer.java#L112 unassigns the parent's label if there is one, but doesn't remove the parent from unassignedProgressOperations , so it's possible for the parent to eventually appear multiple times in unassignedProgressOperations.

**Comment by ljacomet** [comment]
Closing as the fix for this was merged in 7.0
